flowey/flowey_core/src/node.rs

```diff
@@ -251,6 +245,22 @@ impl<T: Serialize + DeserializeOwned> WriteVar<T, VarNotClaimed> {
         WriteVar {
             backing_var: self.backing_var,
             is_secret: self.is_secret,
+            is_side_effect: self.is_side_effect,
+            _kind: std::marker::PhantomData,
+        }
+    }
+}
+
+impl WriteVar<SideEffect, VarNotClaimed> {
+    /// Transforms this writer into one that can be used to write a `T`.
+    ///
+    /// This is useful when a reader only cares about the side effect of an
+    /// operation, but the writer wants to provide output as well.
+    pub fn discard_result<T: Serialize + DeserializeOwned>(self) -> WriteVar<T> {
```

Comment:
why is this needed when the reader can simply `into_side_effect` a `ReadVar<T>`? Having this capability seems spooky, as you've added a runtime failure path in the case where some downstream node attemps to actually read a `T` from this SideEffect var

Reply:
There's no runtime failure path here, since there's no way to get a `ReadVar<T>` from this.

You're right that locally you can just do `into_side_effect` on the read side. But if you're bridging between one caller that wants a side effect (they've given you a `WriteVar<SideEffect>`), and another caller that wants to produce a value (they want a `WriteVar<PathBuf>`), then it's annoying to have to build another step to map between them.

Ultimately I think we should significantly reduce the number of uses of `WriteVar` across the code base (e.g., by having a function-based node that returns a ReadVar<T>), so that perhaps this never gets much use. But for now, the parallelism is useful.

Reply:
(I'll be honest--I decided to add this before I decided to just make `SideEffect` == `()`. But I still think it's a useful tool in the toolbox, more useful than being able to manifest a new `ReadVar` out of an existing `WriteVar`.)

Reply:
Ah, I see. You removed the `new_reader` API entirely. I didn't see that on my first review through. Yeah, that would certainly do the trick to prevent runtime errors.

Yeah, mulling over this API and corresponding implementation - this seems like a reasonable tool in the toolbox. Having to track `is_side_effect` at runtime is a bit unfortunate, but I don't see a way to work around that.

I do think the doc comment could be a bit clearer (e.g: offering an example, along the lines of the one you left in the comment above)

* * *

IIRC, there is code in the interal repo that uses `new_reader`. I'd double check that assertion myself, but as you know, that's no longer an option for me lol.